### ISSUE TITLE
discoverReaders must be called before activateReader

### ISSUE THREAD

**Issue by Nikilaps** [stripe/stripe-terminal-android#246]
I am trying to connect to the Stripe Reader M2 bluetooth reader from my Android app. I am using following sdk version

```
implementation "com.stripe:stripeterminal:2.10.2"
implementation 'com.google.guava:guava:29.0-android'

```
When I am trying to connect with a reader that was discovered using `Terminal.getInstance().discoverReaders()` method, I am getting following exception

```
E/StripeTerminal: discoverReaders must be called before activateReader
    com.stripe.stripeterminal.external.models.TerminalException: discoverReaders must be called before activateReader
        at com.stripe.stripeterminal.resourcerepository.ProxyResourceRepository.activateReader(ProxyResourceRepository.kt:51)
        at com.stripe.stripeterminal.TerminalSession$ConnectReaderOperation.execute(TerminalSession.kt:943)
        at com.stripe.stripeterminal.TerminalSession$ExternalOperation.run$core_publish(TerminalSession.kt:573)
        at com.stripe.stripeterminal.TerminalSession.enqueueOperation$lambda-1(TerminalSession.kt:443)
        at com.stripe.stripeterminal.TerminalSession.$r8$lambda$Ykh0X1lsW7mn-wh2rn1sqLmVjzk(Unknown Source:0)
        at com.stripe.stripeterminal.TerminalSession$$ExternalSyntheticLambda0.run(Unknown Source:4)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:920)
```

This is my related code (removed lots of logging/boilerplate codes for brevity. Other related codes like TerminalApplicationDelegate methods are there, though not mentioned here ). 

```
fun startStripeCardPresent(purchase: Purchase): Boolean {
        AppUtils.log(STRIPE_TAG, "discoverStripeCardReaders")
            if (Terminal.isInitialized()) {
                Terminal.getInstance().clearCachedCredentials()
            } else {
                Terminal.initTerminal(
                    applicationContext,
                    LogLevel.VERBOSE,
                    TokenProvider(cartID = purchase.cart?.id, eventID = eventID),
                    object : TerminalListener {
                        override fun onUnexpectedReaderDisconnect(reader: Reader) {
                             //todo
                        }
                    }
                )
            }
            discoverStripeCardReaders(purchase)
            return true
        }
        return false
}

private fun discoverStripeCardReaders(purchase: Purchase) {
        val config = DiscoveryConfiguration(discoveryMethod = DiscoveryMethod.BLUETOOTH_SCAN, isSimulated = false, timeout = 10)

        var reader: Reader? = null
        Terminal.getInstance().discoverReaders(
            config = config,
            discoveryListener = object : DiscoveryListener {
                override fun onUpdateDiscoveredReaders(readers: List<Reader>) {
                    readers.firstOrNull()?.let {
                        reader = it
                    }
                }
            },
            callback = object: Callback {
                override fun onFailure(e: TerminalException) { }

                override fun onSuccess() {                        reader?.let {                               connectToStripeReader(it, purchase)                        }                        
                }
            }
        )
}

private fun connectToStripeReader(reader: Reader, purchase: Purchase) {        //my backend is not ready, so hard coding location_id for now. This location was created using postman using the Curl request mentioned in documentation
        val connectionConfig = ConnectionConfiguration.BluetoothConnectionConfiguration("tml_Esi2zAqSs0mFBN")
        Terminal.getInstance().connectBluetoothReader(
            reader = reader,
            config = connectionConfig,
            listener = object: BluetoothReaderListener {
                override fun onReportReaderEvent(event: ReaderEvent) {
                }

                override fun onRequestReaderInput(options: ReaderInputOptions) {
                }
            },
            connectionCallback = object: ReaderCallback {
                override fun onSuccess(reader: Reader) {
                    //reader connected
                    AppUtils.log(STRIPE_TAG, "READER CONNECTED : ${reader.toString()}")
                }

                override fun onFailure(e: TerminalException) {
                    AppUtils.log(STRIPE_TAG, "READER FAILED TO CONNECT : ${e.localizedMessage}") // This is where I am getting error mentioned above.
                    onSwiperDisconnected()
                }
            }
        )
}
```

Basically, I am discovering my reader alright (I only have one reader hardware). When discovery is complete (I am not trying to cancel discovery when a reader is found), I tried to connect with the reader by hardcoding a location ID that was created using Postman (my backend is not ready yet). 

Any idea?

**Comment by chenmin-stripe** [comment]
@Nikilaps 

Could you share your reader serial ID?

**Comment by Nikilaps** [comment]
@chenmin-stripe  
They are STRM26138021154 and STRM26138021159.

Is there any serial number specific thing that we need to do? My client has their own readers with different serial numbers. Or is this some location specific blocking?

**Comment by Nikilaps** [comment]
@chenmin-stripe Any updates on this? Do you want any other information?

**Comment by chenmin-stripe** [comment]
Hi @Nikilaps 
You should call `connectBluetoothReader` upon `onUpdateDiscoveredReaders`. I saw you put it inside `callback`, that one is used to receive messages whether connect is success or not.
```
private fun discoverStripeCardReaders(purchase: Purchase) {
        val config = DiscoveryConfiguration(discoveryMethod = DiscoveryMethod.BLUETOOTH_SCAN, isSimulated = false, timeout = 10)

        var reader: Reader? = null
        Terminal.getInstance().discoverReaders(
            config = config,
            discoveryListener = object : DiscoveryListener {
                override fun onUpdateDiscoveredReaders(readers: List<Reader>) {
                    readers.firstOrNull()?.let {
                        connectToStripeReader(it, purchase) 
                    }
                }
            },
            callback = object: Callback {
                override fun onFailure(e: TerminalException) { }

                override fun onSuccess() {                       

                }
            }
        )
}

```

[doc](https://stripe.com/docs/terminal/payments/connect-reader?terminal-sdk-platform=android&reader-type=simulated) for your reference

**Comment by Nikilaps** [comment]
Thanks, Its working.
